src/10-pipe-maze.cpp

```diff
@@ -0,0 +1,208 @@
+#include <iostream>
+#include <exception>
+#include <string>
+#include <vector>
+
+#include "10-pipe-maze.h"
+
+using namespace std;
+
+class PipePath
+{
+    private:
+    const vector<string> m;
+    const int xLen, yLen;
+    vector<string> pm;
+    int x, y;
+    int prevX, prevY;
+    int stepCount;
+
+    public:
+    PipePath(vector<string> _m) : m(_m), xLen(m[0].size()), yLen(m.size())
+    {
+        pm = vector<string>(yLen, string(xLen, '.'));
+        stepCount = 0;
+        prevX = -1;
+        prevY = -1;
+
+        moveToStartingPosition();
+        discoverPath();
+    }
+
+    void print()
+    {
+        for (int line = 0; line < yLen; line++) {
+            cout << m[line] << " > ";
+            for (int i = 0; i < m[line].size(); i++) {
+                if (pm[line][i] == '*') cout << m[line][i];
+                else cout << pm[line][i];
+            }
+            cout << endl;
+        }
+    }
+
+    int countStepsToFarthestPoint()
```

Comment:
I just learned that you can tag a method as `const` to indicate that it doesn't mutate any state in the object itself.

```cpp
int countStepsToFarthestPoint() const {
    return stepCount / 2;
}
```